src/Api/Rest/Rest.php

```diff
@@ -0,0 +1,225 @@
+//		}
+
+		// Get the Profile object for the webservice requested.
+		$this->profile = Factory::getProfile($this->getContainer(), $clientName, $resourceName, $version, $operation);
+
+		$this->webservice = Factory::getWebservice($this->getContainer(), $operation, $this->getOptions());
+
+		// Set initial status code to OK.
+		$this->setStatusCode(200);
+
+		// We do not want some unwanted text to appear before output.
+		ob_start();
+
+		try
+		{
+			// Execute the web service operation.
+			$this->resource = $this->webservice->execute($this->profile);
+
+			$executionErrors = ob_get_contents();
+			ob_end_clean();
+//			ob_end_flush();		// TEMPORARY so we can see any error messages.
+		}
+		catch (\Exception $e)
+		{
+			$executionErrors = ob_get_contents();
+			ob_end_clean();
+
+			throw $e;
+		}
+
+		$messages = $this->app->getMessageQueue();
+
+		if (!empty($executionErrors))
+		{
+			$messages[] = array('message' => $executionErrors, 'type' => 'notice');
+		}
+
+		if (!empty($messages))
+		{
+			// If we are not in debug mode we will take out everything except errors.
+			if ($this->app->get('webservices.debug_webservices', 0) == 0)
+			{
+				foreach ($messages as $key => $message)
+				{
+					if ($message['type'] != 'error')
+					{
+						unset($messages[$key]);
+					}
+				}
+			}
+
+			$this->resource->setData('_messages', $messages);
+		}
+
+		return $this;
+	}
+
+	/**
+	 * Method to send the application response to the client.
+	 * All headers will be sent prior to the main application output data.
+	 *
+	 * @return  void
+	 *
+	 * @since   1.2
+	 */
+	public function render()
+	{
+		// Set token to uri if used in that way
+		$token = $this->getOptions()->get('accessToken', '');
+		$client = $this->getOptions()->get('webserviceClient', '');
+
+		if (!empty($token))
+		{
+			$this->webservice->setUriParams($this->app->get('webservices.oauth2_token_param_name', 'access_token'), $token);
+		}
+
+		if ($client == 'administrator')
+		{
+			$this->webservice->setUriParams('webserviceClient', $client);
+		}
+
+		// Calculate runtime to this point.
+		$runtime = microtime(true) - $this->app->startTime;
+
+		// Set headers.
+		$this->app->setHeader('Status', $this->webservice->statusCode . ' ' . $this->webservice->statusText, true);
+		$this->app->setHeader('Server', '', true);
+		$this->app->setHeader('X-Runtime', $runtime, true);
+		$this->app->setHeader('Access-Control-Allow-Origin', '*', true);
+		$this->app->setHeader('Pragma', 'public', true);
+		$this->app->setHeader('Expires', '0', true);
+		$this->app->setHeader('Cache-Control', 'must-revalidate, post-check=0, pre-check=0', true);
+		$this->app->setHeader('Cache-Control', 'private', false);
+		$this->app->setHeader('Webservice-Name', $this->webservice->webserviceName, true);
+		$this->app->setHeader('Webservice-Version', $this->webservice->webserviceVersion, true);
+
+		// Push results into the document.
+		if ($this->resource instanceof Resource)
+		{
+			$this->app->setBody(
```

Comment:
Yes, this is one area that needs particular attention.  Something for the next round of refactoring.